priv/perf/apps/load_test/lib/ethereum/ethereum.ex

```diff
@@ -108,6 +120,232 @@ defmodule LoadTest.Ethereum do
+    end
+  end
+
+  defp execute_submit_typed(typed_data_signed) do
+    WatcherInfoAPI.Api.Transaction.submit_typed(LoadTest.Connection.WatcherInfo.client(), typed_data_signed)
+  end
+
+  defp process_transaction_result(result, amount_in_wei, input_address, output_address, currency, tries) do
+    case {result, tries} do
+      {%{"code" => "create:client_error"}, 0} ->
+        {:error, result}
+
+      {%{
+         "result" => "complete",
+         "transactions" => [
+           %{
+             "sign_hash" => sign_hash,
+             "typed_data" => typed_data,
+             "txbytes" => txbytes
+           }
+         ]
+       }, _} ->
+        {:ok, [sign_hash, typed_data, txbytes]}
+
+      _ ->
+        Process.sleep(1_000)
+        create_transaction(amount_in_wei, input_address, output_address, currency, tries - 1)
+    end
+  end
+
+  defp root_chain_get_eth_balance(address, 0) do
+    raise "failed to fetch root chain balance for #{inspect(address)}"
+  end
+
+  defp root_chain_get_eth_balance(address, counter) do
+    response = eth_account_get_balance(address)
+
+    case response do
+      {:ok, initial_balance} ->
+        {initial_balance, ""} = initial_balance |> String.replace_prefix("0x", "") |> Integer.parse(16)
+        initial_balance
+
+      _ ->
+        Process.sleep(1_000)
+        root_chain_get_eth_balance(address, counter - 1)
+    end
+  end
+
+  defp eth_account_get_balance(address) do
+    Ethereumex.HttpClient.eth_get_balance(address)
+  end
+
+  defp root_chain_get_erc20_balance(address, currency, 0) do
+    do_root_chain_get_erc20_balance(address, currency)
+  end
+
+  defp root_chain_get_erc20_balance(address, currency, counter) do
+    case do_root_chain_get_erc20_balance(address, currency) do
+      {:ok, balance} ->
+        balance
+
+      _ ->
+        Process.sleep(1_000)
+        root_chain_get_erc20_balance(address, currency, counter - 1)
+    end
+  end
+
+  defp do_root_chain_get_erc20_balance(address, currency) do
+    data = ABI.encode("balanceOf(address)", [Encoding.to_binary(address)])
+
+    case Ethereumex.HttpClient.eth_call(%{to: Encoding.to_hex(currency), data: Encoding.to_hex(data)}) do
+      {:ok, result} ->
+        balance =
+          result
+          |> Encoding.to_binary()
+          |> ABI.TypeDecoder.decode([{:uint, 256}])
+          |> hd()
+
+        {:ok, balance}
+
+      error ->
+        error
+    end
+  end
+
+  defp fetch_balance(address, amount, currency, counter) do
+    response =
+      case account_get_balances(address) do
+        {:ok, response} ->
+          decoded_response = Jason.decode!(response.body)
+          Enum.find(decoded_response["data"], fn data -> data["currency"] == currency end)
+
+        _ ->
+          # socket closed etc.
+          :error
+      end
+
+    case response do
+      # empty response is considered no account balance!
+      nil when amount == 0 ->
+        nil
+
+      %{"amount" => ^amount} = balance ->
```

Comment:
you're pinning the amount twice. Once in the calling function `fetch_childchain_balance` and the second time here. is this necessary?  what are you trying to acomplish?

I see you're also mixing approaches. 
https://github.com/omgnetwork/elixir-omg/pull/1745/files#diff-3e3ab486b14889eab1712a81574bd2c6a194527da3c65eec103b399e86a6a896R125

perhaps this should be `do_fetch_balance `. I've seen you used this naming convention before.

Reply:
Now I use `Sync` module for every retry in the perf project